src/Microsoft.Sbom.Parsers.Spdx30SbomParser/Parser/SPDX30Parser.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.Sbom.Contracts;
+using Microsoft.Sbom.Extensions.Entities;
+using Microsoft.Sbom.JsonAsynchronousNodeKit;
+using Microsoft.Sbom.JsonAsynchronousNodeKit.Exceptions;
+using Microsoft.Sbom.Parsers.Spdx30SbomParser.Entities;
+using SPDXConstants = Microsoft.Sbom.Parsers.Spdx30SbomParser.Constants;
+
+namespace Microsoft.Sbom.Parser;
+
+#nullable enable
+/// <summary>
+/// A parser for SPDX 3.0 SBOMs.
+/// </summary>
+/// <remarks>
+/// This class is not Thread-safe since the stream and JsonReaders assume a single forward-only reader.
+/// Because of the use of recursion in <see cref="LargeJsonParser30"/>, this class is also not suitable for parsing deeply nested json objects.
+/// </remarks>
+public class SPDX30Parser : ISbomParser
+{
+    public const string ContextProperty = SPDXConstants.SPDXContextHeaderName;
+    public const string GraphProperty = SPDXConstants.SPDXGraphHeaderName;
+    public static readonly IReadOnlyCollection<string> RequiredFields = new List<string>
+    {
+        ContextProperty,
+        GraphProperty,
+    };
+
+    public string? RequiredComplianceStandard;
+    public IReadOnlyCollection<string>? EntitiesToEnforceComplianceStandardsFor;
+    public SpdxMetadata Metadata = new SpdxMetadata();
+    private readonly LargeJsonParser parser;
+    private readonly IList<string> observedFieldNames = new List<string>();
+    private readonly bool requiredFieldsCheck = true;
+    private readonly JsonSerializerOptions jsonSerializerOptions;
+    private bool parsingComplete = false;
+    private readonly ManifestInfo spdxManifestInfo = new()
+    {
+        Name = SPDXConstants.SPDXName,
+        Version = SPDXConstants.SPDXVersion,
+    };
+
+    private readonly IReadOnlyCollection<string> entitiesWithDifferentNTIARequirements = new List<string>
+    {
+        "SpdxDocument",
+        "File",
+        "Package",
+    };
+
+    public SPDX30Parser(
+        Stream stream,
+        string? requiredComplianceStandard = null,
+        JsonSerializerOptions? jsonSerializerOptions = null,
+        int? bufferSize = null)
+    {
+        this.jsonSerializerOptions = jsonSerializerOptions ?? new JsonSerializerOptions
+        {
+            Converters = { new ElementSerializer() },
+        };
+
+        var handlers = new Dictionary<string, PropertyHandler>
+        {
+            { ContextProperty, new PropertyHandler<string>(ParameterType.Array) },
+            { GraphProperty, new PropertyHandler<JsonNode>(ParameterType.Array) },
+        };
+
+        switch (requiredComplianceStandard)
+        {
+            case "NTIA":
+                this.EntitiesToEnforceComplianceStandardsFor = this.entitiesWithDifferentNTIARequirements;
+                this.RequiredComplianceStandard = requiredComplianceStandard;
+                break;
```

Comment:
What do we want the behavior to be when an invalid requiredComplianceStandard value is provided? It seems like we're just ignoring it right now but I would think that we would want to throw

Reply:
Yes, throwing a parser exception would be right. I throw it later in the parsing process so it would've eventually thrown but adding it here would be better and more readable.